src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                        editor.ReplaceNode(match, replacement);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Goes through the parameters of the original method to get the location that the parameter
+        /// and argument should be introduced.
+        /// </summary>
+        private static int GetInsertionIndex(Compilation compilation, IMethodSymbol methodSymbol,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            Contract.ThrowIfNull(parameterList);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in methodSymbol.Parameters)
+            {
+                // Want to skip optional parameters, params parameters, and CancellationToken since they should be at
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// For the trampoline case, it goes through the invocations and adds an argument which is a 
+        /// call to the extracted method.
+        /// Introduces a new method overload or new trampoline method.
+        /// Updates the original method site with a newly introduced parameter.
+        /// </summary>
+        private static async Task<Solution> ModifyDocumentInvocationsTrampolineAndIntroduceParameterAsync(
+            Compilation compilation, Solution modifiedSolution, Document currentDocument, Document originalDocument,
+            ImmutableArray<TInvocationExpressionSyntax> invocations, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            bool allOccurrences, string parameterName, TExpressionSyntax expression, bool trampoline, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(currentDocument);
+            var semanticFacts = currentDocument.GetRequiredLanguageService<ISemanticFactsService>();
+            var invocationSemanticModel = await currentDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var root = await currentDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = currentDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
```

Comment:
this name seems like it's going to be weird (And not .net-y).  It will be something like `Foo_bar`.  Why not just `GetBar()`?

Reply:
still applies.

Reply:
fixed